bbPress KSES

johngodley opened the pull request:
Allows block markup to survive bbPress, as well as maps Gutenberg blocks to a set of KSES rules. Together this will allow a non-administrator account to post content without it being mangled in some way.

Additionally the following Gutenberg features have been disabled as they're not needed in the situations this plugin runs in:
- typography controls
- colours
- drop caps
- custom class names
- anchors

Based on code by @dd32

johngodley committed:
Add function to map allowed blocks to KSES

Changes to Gutenberg could break this.

johngodley committed:
Remove typography controls

We don’t need these on the places this plugin runs

johngodley committed:
Remove block support for advanced features

Remove support for custom class names, anchors, html, and color choices. It’s unlikely these will be needed where this plugin is used

johngodley committed:
Allow block markup in bbPress

Allows block markup in bbPress and then connects up the KSES filters to a set that handle the core blocks

johngodley commented:
@dd32, I've gone with your code and mapped it a bit further onto bbPress. I need to give it a proper test but it seems like a good starting point.

If Gutenberg markup changes then the rules are fragile, and could filter stuff out (resulting in validation errors when edited). For the most part we'll just need to keep things up to date.

Class names could change more often than anything else. This PR locks things down to specific class names, but opening them up to any class name would make it less likely to break. Do you think this would be a good idea? There's the potential for mischief (people sneaking in classes that apply unwanted styles), but I don't know if this is a big enough problem to worry about.

dd32 commented:
My thoughts:
 - We probably don't need to sanitize the tag classes, it seems like it's a very small edge-case. Especially given how fragile it'll be with Gutenberg changes. [bbPress doesn't do it either, allowing any class on `<pre>` tags](https://github.com/bbpress/bbPress/blob/394ae31c059228e866240076db2dc2ca2ddc8a7c/src/includes/common/formatting.php#L45). The main thing needing to be protected against is malicious tags, and tags that break DOM structure.. CSS will just break the visual styles.  Removing classes from the tags will break the Block upon edit, so that's another reason that I don't think the potential UX breakage is worth the extra-safe kses.
 - I had updated the Block HTML Comments regex to `~&lt;!-- wp:(.+?)--&gt;~i` but it's probably not really needed, and I'm not sure if custom blocks have that prefix..

Other than that, this looks much more complete than my initial code, and much better for release.

johngodley committed:
Only modify KSES if no unfiltered_html

Admins can already post blocks

johngodley committed:
Tweak the comment regex

Specifically match the `wp:paragraph` format, while allowing custom blocks

johngodley committed:
Allow class names in all blocks

There’s a small chance for mischief, but it helps prevent fragility when Gutenberg changes class names

johngodley committed:
Mirror readme.txt in readme.md

So people visitng on Github see the same info

johngodley commented:
Great. I've opened up the `class` filter. I like tightening up the comment regex too. Custom blocks do have a different namespace so I've modified it to accept those. Thanks Dion!

johngodley committed:
Bump versions
